core/drivers/scmi-msg/shm_msg.c

```diff
@@ -0,0 +1,93 @@
+// SPDX-License-Identifier: BSD-2-Clause
+/*
+ * Copyright (c) 2021, Linaro Limited
+ */
+#include <assert.h>
+#include <drivers/scmi-msg.h>
+#include <drivers/scmi.h>
+#include <io.h>
+#include <kernel/misc.h>
+#include <stdint.h>
+#include <string.h>
+#include <trace.h>
+#include <util.h>
+
+#include "common.h"
+
+/**
+ * struct msg_header - MSG formatted header for MSG base shared memory transfer
+ *
+ * @message_header: 32bit header used in MSG shared memory protocol
+ * @payload: SCMI message payload data
+ */
+struct msg_header {
+	uint32_t message_header;
+	uint32_t payload[];
+};
+
+/* Bit fields packed in msg_header::message_header */
+#define MSG_ID_MASK		GENMASK_32(7, 0)
+#define MSG_ID(_hdr)		((_hdr) & MSG_ID_MASK)
+
+#define MSG_TYPE_MASK		GENMASK_32(9, 8)
+#define MSG_TYPE(_hdr)		(((_hdr) & MSG_TYPE_MASK) >> 8)
+
+#define MSG_PROT_ID_MASK	GENMASK_32(17, 10)
+#define MSG_PROT_ID(_hdr)	(((_hdr) & MSG_PROT_ID_MASK) >> 10)
+
+/*
+ * Creates a SCMI message instance in secure memory and push it in the SCMI
+ * message drivers. Message structure contains SCMI protocol meta-data and
+ * references to input payload in secure memory and output message buffer
+ * in shared memory.
+ */
+TEE_Result scmi_entry_msg(unsigned int channel_id, void *in_buf, size_t in_size,
+			  void *out_buf, size_t *out_size, uint32_t *sec_buf)
+{
+	struct scmi_msg_channel *channel = plat_scmi_get_channel(channel_id);
+	struct msg_header *hdr = NULL;
+	struct scmi_msg msg = { };
+	uint32_t msg_header = 0;
+
+	if (!channel)
+		return TEE_ERROR_BAD_PARAMETERS;
+
+	if (!channel || !in_buf || in_size < sizeof(struct msg_header) ||
```

Comment:
`in_buf` and `out_buf` are checked against `NULL` in `cmd_process_msg_channel()` too.